pkg/git/client.go

```diff
@@ -133,6 +143,14 @@ func (c *client) Clone(ctx context.Context, repoID, remote, branch, destination
 		)
 	)
 
+	authArgs := []string{}
+	if c.username != "" && c.password != "" {
+		token := fmt.Sprintf("%s:%s", c.username, c.password)
+		encodedToken := base64.StdEncoding.EncodeToString([]byte(token))
+		header := fmt.Sprintf("Authorization: Basic %s", encodedToken)
+		authArgs = append(authArgs, "-c", fmt.Sprintf("http.extraHeader=%s", header))
+	}
+
```

Comment:
At first, thank you for investigating and the fix 🙏 I don't know such an option! It's so helpful 👍

I tried the expected command with a private repo `git -c http.extraHeader="Authorization: Basic  <base64 encoded value of username:password>" clone https://github.com/ffjlabo-playground/git-test.git` on mac. 
I encountered the behaviors written below. Does below also reproduce on your machine?

- It shows the prompts for username and password like this↓
```console
% git -c http.extraHeader="Authorization: Basic <base64 encoded value of username:password>" clone https://github.com/ffjlabo-playground/git-test.git
Cloning into 'git-test'...
Username for 'https://github.com':
```

- When I input the username and password as empty, then It shows the auth error like this↓
```console
% git -c http.extraHeader="Authorization: Basic <base64 encoded value of username:password>" clone https://github.com/ffjlabo-playground/git-test.git
Cloning into 'git-test'...
Username for 'https://github.com':
Password for 'https://github.com':
remote: Support for password authentication was removed on August 13, 2021.
remote: Please see https://docs.github.com/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.
fatal: Authentication failed for 'https://github.com/ffjlabo-playground/git-test.git/'
```

Reply:
Sorry...
It worked when we looked into it before, but we will look into it again and verify.

Reply:
Thank you! Don't worry! I think this is a challenging. I'm so helpful.

Reply:
Hi @sZma5a ! Do you have any updates on the above? If you want some helps, feel free to ping me 👍

Reply:
Sorry for the delay.
Regarding git, I was able to perform a clone using PAT in my environment with the following command.

```
git -c http.extraHeader="Authorization: Basic BASE64_PAT" clone --mirror https://github.com/PRIVATE_REPO ./tmp2 
```

In addition, the Golang code that was tested is shown below.

```
package main

import (
	"context"
	"fmt"

	"github.com/pipe-cd/pipecd/pkg/git"
)

func main() {
	uo := git.WithUserName("USERNAME")
	po := git.WithPassword("PAT")
	c, err := git.NewClient(uo, po)
	if err != nil {
		fmt.Println(err)
		panic(err)
	}
	ctx := context.Background()
	c.Clone(
		ctx,
		"test1",
		"https://github.com/PRIVATE_REPO",
		"main",
		"./tmp",
	)
}

```

Reply:
@ffjlabo 
Sorry for the delay.
Regarding git, I was able to perform a clone using PAT in my environment with the following command.

```
git -c http.extraHeader="Authorization: Basic {base64(username:token)}" clone --mirror https://github.com/PRIVATE_REPO ./tmp2 
```

In addition, the Golang code that was tested is shown below.

```
package main

import (
	"context"
	"fmt"

	"github.com/pipe-cd/pipecd/pkg/git"
)

func main() {
	uo := git.WithUserName("USERNAME")
	po := git.WithPassword("PAT")
	c, err := git.NewClient(uo, po)
	if err != nil {
		fmt.Println(err)
		panic(err)
	}
	ctx := context.Background()
	c.Clone(
		ctx,
		"test1",
		"https://github.com/PRIVATE_REPO",
		"main",
		"./tmp",
	)
}

```

Reply:
I checked with the below command and succeeded in cloning the repository.
```
git -c http.extraHeader="Authorization: Basic $(echo -n "x-access-token:$(gh auth token)" | base64)" clone --mirror https://github.com/Warashi/PRIVATE.git ./tmp2
```

When I removed the `-n` option from echo as below, I got the same error with @ffjlabo.
```
git -c http.extraHeader="Authorization: Basic $(echo "x-access-token:$(gh auth token)" | base64)" clone --mirror https://github.com/Warashi/PRIVATE.git ./tmp2
```

So, I think that @ffjlabo appended the `\n` to the password before encoding it.

Reply:
Sorry to be late 🙏 
It's my fault, as Wrashi-san says. The command worked as expected.
Thank you!

BTW, there are some fixed points in this comment, so plz check it🙏  
https://github.com/pipe-cd/pipecd/pull/4571#pullrequestreview-2159787547